File: scripts/apply-migrations.js
/**
 * Apply Database Migrations Programmatically
 * 
 * This script reads all migration files and applies them to the Supabase database.
 * 
 * USAGE:
 *   node scripts/apply-migrations.js
 * 
 * REQUIREMENTS:
 *   - SERVICE_ROLE_KEY environment variable must be set
 *   - Or you can pass it as the first argument
 * 
 * EXAMPLE:
 *   SERVICE_ROLE_KEY=your-service-role-key node scripts/apply-migrations.js
 *   OR
 *   node scripts/apply-migrations.js your-service-role-key
 */

const fs = require('fs');
const path = require('path');
const https = require('https');

const SUPABASE_URL = 'https://qvslbiceoonrgjxzkotb.supabase.co';
const SERVICE_ROLE_KEY = process.argv[2] || process.env.SERVICE_ROLE_KEY;

if (!SERVICE_ROLE_KEY) {
  console.error('ERROR: SERVICE_ROLE_KEY is required');
  console.error('');
  console.error('Get your Service Role Key from:');
  console.error('https://supabase.com/dashboard/project/qvslbiceoonrgjxzkotb/settings/api');
  console.error('');
  console.error('Usage:');
  console.error('  SERVICE_ROLE_KEY=your-key node scripts/apply-migrations.js');
  console.error('  OR');
  console.error('  node scripts/apply-migrations.js your-key');
  process.exit(1);
}

async function executeSQL(sql) {
  return new Promise((resolve, reject) => {
    const postData = JSON.stringify({ query: sql });
    
    const options = {
      hostname: 'qvslbiceoonrgjxzkotb.supabase.co',
      port: 443,
      path: '/rest/v1/rpc/exec_sql',
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': Buffer.byteLength(postData),
        'apikey': SERVICE_ROLE_KEY,
        'Authorization': `Bearer ${SERVICE_ROLE_KEY}`
      }
    };

    const req = https.request(options, (res) => {
      let data = '';
      
      res.on('data', (chunk) => {
        data += chunk;
      });
      
      res.on('end', () => {
        if (res.statusCode >= 200 && res.statusCode < 300) {
          resolve({ success: true, data });
        } else {
          reject({ success: false, statusCode: res.statusCode, data });
        }
      });
    });

    req.on('error', (error) => {
      reject(error);
    });

    req.write(postData);
    req.end();
  });
}

async function applyMigrations() {
  console.log('🔧 Applying Database Migrations...\n');
  
  const migrationsDir = path.join(__dirname, '..', 'supabase', 'migrations');
  const migrationFiles = fs.readdirSync(migrationsDir)
    .filter(f => f.endsWith('.sql'))
    .sort();
  
  console.log(`Found ${migrationFiles.length} migration files\n`);
  
  for (const file of migrationFiles) {
    const filePath = path.join(migrationsDir, file);
    const sql = fs.readFileSync(filePath, 'utf8');
    
    console.log(`Applying: ${file}...`);
    
    try {
      // Note: Supabase doesn't have exec_sql RPC by default
      // We need to use SQL Editor or psql directly
      console.log(`  ⚠️  Skipped (requires SQL Editor or psql)`);
    } catch (error) {
      console.error(`  ❌ Failed: ${error.message}`);
      throw error;
    }
  }
  
  console.log('\n✅ Migration process complete');
  console.log('\nNOTE: Migrations must be applied via:');
  console.log('  1. Supabase Dashboard SQL Editor');
  console.log('  2. psql with database connection string');
  console.log('  3. See MIGRATION_GUIDE.md for details');
}

applyMigrations().catch(console.error);

File: scripts/apply-migrations.py
#!/usr/bin/env python3
"""
Apply Supabase Migrations using Direct Database Connection

This script applies all migration files to your Supabase PostgreSQL database.

USAGE:
    python3 scripts/apply-migrations.py <DATABASE_URL>

EXAMPLE:
    python3 scripts/apply-migrations.py "postgresql://postgres:[PASSWORD]@db.[YOUR-PROJECT-ID].supabase.co:5432/postgres"

Get your database URL from:
    https://supabase.com/dashboard/project/[YOUR-PROJECT-ID]/settings/database
    Look for "Connection string" under "Connection pooling"
"""

import sys
import os
import glob
from pathlib import Path

try:
    import psycopg2
except ImportError:
    print("ERROR: psycopg2 not installed")
    print("Install it with: pip install psycopg2-binary")
    sys.exit(1)

def apply_migrations(database_url):
    """Apply all migrations to the database"""
    
    # Connect to database
    print("🔗 Connecting to database...")
    try:
        conn = psycopg2.connect(database_url)
        conn.autocommit = False
        cursor = conn.cursor()
        print("✓ Connected successfully\n")
    except Exception as e:
        print(f"❌ Connection failed: {e}")
        sys.exit(1)
    
    # Get migration files
    migrations_dir = Path(__file__).parent.parent / 'supabase' / 'migrations'
    migration_files = sorted(glob.glob(str(migrations_dir / '*.sql')))
    
    print(f"📁 Found {len(migration_files)} migration files\n")
    
    # Apply each migration
    for migration_file in migration_files:
        filename = os.path.basename(migration_file)
        print(f"  Applying: {filename}...")
        
        try:
            with open(migration_file, 'r') as f:
                sql_content = f.read()
            
            cursor.execute(sql_content)
            conn.commit()
            print("    ✓ Success")
        except Exception as e:
            print(f"    ❌ Failed: {e}")
            conn.rollback()
            print("\n⚠️  Migration failed. Rolling back...")
            sys.exit(1)
    
    # Apply seed data
    seed_file = Path(__file__).parent.parent / 'supabase' / 'seed.sql'
    if seed_file.exists():
        print("\n🌱 Applying seed data...")
        try:
            with open(seed_file, 'r') as f:
                sql_content = f.read()
            
            cursor.execute(sql_content)
            conn.commit()
            print("  ✓ Seed data applied")
        except Exception as e:
            print(f"  ❌ Seed data failed: {e}")
            conn.rollback()
    
    cursor.close()
    conn.close()
    
    print("\n✅ All migrations applied successfully!")
    print("\nNext steps:")
    print("  1. Verify tables exist")
    print("  2. Run Phase 1 validation: ./scripts/validate-phase-1.sh")

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print(__doc__)
        sys.exit(1)
    
    database_url = sys.argv[1]
    apply_migrations(database_url)
